tests/test_qaoa.py

```diff
@@ -199,3 +200,105 @@ def test_maxcut_output(self, graph, target_hamiltonian):
         assert cost_coeffs == target_coeffs
         assert cost_ops == target_ops
         assert cost_wires == target_wires
+
+
+class TestUtils:
+    """Tests that the utility functions are working properly"""
+
+    @pytest.mark.parametrize(
+        ("hamiltonian", "value"),
+        (
+            (qml.Hamiltonian([1, 1], [qml.PauliZ(0), qml.PauliZ(1)]), True),
+            (qml.Hamiltonian([1, 1], [qml.PauliX(0), qml.PauliZ(1)]), False),
+            (qml.Hamiltonian([1, 1], [qml.PauliZ(0) @ qml.Identity(1), qml.PauliZ(1)]), True),
+            (qml.Hamiltonian([1, 1], [qml.PauliZ(0), qml.PauliX(0) @ qml.PauliZ(1)]), False),
+        ),
+    )
+    def test_diagonal_terms(self, hamiltonian, value):
+        assert qaoa.layers._diagonal_terms(hamiltonian) == value
+
+
+class TestLayers:
+    """Tests that the cost and mixer layers are being constructed properly"""
+
+    def test_mixer_layer_errors(self):
+        """Tests that the mixer layer is throwing the correct errors"""
+
+        hamiltonian = [[1, 1], [1, 1]]
+
+        with pytest.raises(ValueError, match=r"hamiltonian must be of type pennylane.Hamiltonian"):
+            qaoa.mixer_layer(0.1, hamiltonian)
+
+    def test_cost_layer_errors(self):
+        """Tests that the cost layer is throwing the correct errors"""
+
+        hamiltonian = [[1, 1], [1, 1]]
+
+        with pytest.raises(ValueError, match=r"hamiltonian must be of type pennylane.Hamiltonian"):
+            qaoa.cost_layer(0.1, hamiltonian)
+
+        hamiltonian = qml.Hamiltonian([1, 1], [qml.PauliZ(0), qml.PauliX(1)])
+
+        with pytest.raises(ValueError, match=r"hamiltonian must be written only in terms of PauliZ and Identity gates"):
+            qaoa.cost_layer(0.1, hamiltonian)
+
+    @pytest.mark.parametrize(
+        ("mixer", "gates"),
+        [
+            [
+                qml.Hamiltonian([1, 1], [qml.PauliX(0), qml.PauliX(1)]),
+                [qml.PauliRot(2, "X", wires=[0]), qml.PauliRot(2, "X", wires=[1])]
+            ],
+            [
+                qaoa.xy_mixer(Graph([(0, 1), (1, 2), (2, 0)])),
+                [qml.PauliRot(1, "XX", wires=[0, 1]), qml.PauliRot(1, "YY", wires=[0, 1]),
```

Comment:
Nice, I never saw this class before! Maybe we need to add it to https://pennylane.readthedocs.io/en/latest/introduction/operations.html

Reply:
Yes, definitely! `MultiRZ` should probably be added as well

Reply:
@trbromley @Lucaman99 Let's make sure that the documentation is fully up to date next week during the feature freeze :)

Reply:
@josh146 Sounds good!